// cypress-svelte-unit-test/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.polyfillFetchIfNeeded = exports.injectStylesBeforeElement = exports.cleanupStyles = exports.checkMountModeEnabled = void 0;
var unfetch_1 = __importDefault(require("unfetch"));
function checkMountModeEnabled() {
    // @ts-ignore
    if (Cypress.spec.specType !== 'component') {
        throw new Error("In order to use mount or unmount functions please place the spec in component folder");
    }
}
exports.checkMountModeEnabled = checkMountModeEnabled;
/**
 * Remove any style or extra link elements from the iframe placeholder
 * left from any previous test
 */
function cleanupStyles(document) {
    var styles = document.body.querySelectorAll('style');
    styles.forEach(function (styleElement) {
        document.body.removeChild(styleElement);
    });
    var links = document.body.querySelectorAll('link[rel=stylesheet]');
    links.forEach(function (link) {
        document.body.removeChild(link);
    });
}
exports.cleanupStyles = cleanupStyles;
/**
 * Insert links to external style resources.
 */
function insertStylesheets(stylesheets, document, el) {
    stylesheets.forEach(function (href) {
        var link = document.createElement('link');
        link.type = 'text/css';
        link.rel = 'stylesheet';
        link.href = href;
        document.body.insertBefore(link, el);
    });
}
/**
 * Inserts a single stylesheet element
 */
function insertStyles(styles, document, el) {
    styles.forEach(function (style) {
        var styleElement = document.createElement('style');
        styleElement.appendChild(document.createTextNode(style));
        document.body.insertBefore(styleElement, el);
    });
}
function insertSingleCssFile(cssFilename, document, el, log) {
    return cy.readFile(cssFilename, { log: log }).then(function (css) {
        var style = document.createElement('style');
        style.appendChild(document.createTextNode(css));
        document.body.insertBefore(style, el);
    });
}
/**
 * Reads the given CSS file from local file system
 * and adds the loaded style text as an element.
 */
function insertLocalCssFiles(cssFilenames, document, el, log) {
    return Cypress.Promise.mapSeries(cssFilenames, function (cssFilename) {
        return insertSingleCssFile(cssFilename, document, el, log);
    });
}
/**
 * Injects custom style text or CSS file or 3rd party style resources
 * into the given document.
 */
exports.injectStylesBeforeElement = function (options, document, el) {
    // first insert all stylesheets as Link elements
    var stylesheets = [];
    if (typeof options.stylesheet === 'string') {
        stylesheets.push(options.stylesheet);
    }
    else if (Array.isArray(options.stylesheet)) {
        stylesheets = stylesheets.concat(options.stylesheet);
    }
    if (typeof options.stylesheets === 'string') {
        options.stylesheets = [options.stylesheets];
    }
    if (options.stylesheets) {
        stylesheets = stylesheets.concat(options.stylesheets);
    }
    insertStylesheets(stylesheets, document, el);
    // insert any styles as <style>...</style> elements
    var styles = [];
    if (typeof options.style === 'string') {
        styles.push(options.style);
    }
    else if (Array.isArray(options.style)) {
        styles = styles.concat(options.style);
    }
    if (typeof options.styles === 'string') {
        styles.push(options.styles);
    }
    else if (Array.isArray(options.styles)) {
        styles = styles.concat(options.styles);
    }
    insertStyles(styles, document, el);
    // now load any css files by path and add their content
    // as <style>...</style> elements
    var cssFiles = [];
    if (typeof options.cssFile === 'string') {
        cssFiles.push(options.cssFile);
    }
    else if (Array.isArray(options.cssFile)) {
        cssFiles = cssFiles.concat(options.cssFile);
    }
    if (typeof options.cssFiles === 'string') {
        cssFiles.push(options.cssFiles);
    }
    else if (Array.isArray(options.cssFiles)) {
        cssFiles = cssFiles.concat(options.cssFiles);
    }
    return insertLocalCssFiles(cssFiles, document, el, options.log);
};
/**
 * Replaces window.fetch with a polyfill based on XMLHttpRequest
 * that Cypress can spy on and stub
 * @see https://www.cypress.io/blog/2020/06/29/experimental-fetch-polyfill/
 */
function polyfillFetchIfNeeded() {
    // @ts-ignore
    if (Cypress.config('experimentalFetchPolyfill')) {
        // @ts-ignore
        if (!cy.state('fetchPolyfilled')) {
            delete window.fetch;
            window.fetch = unfetch_1.default;
            // @ts-ignore
            cy.state('fetchPolyfilled', true);
        }
    }
}
exports.polyfillFetchIfNeeded = polyfillFetchIfNeeded;
